cranelift/codegen/src/isa/aarch64/lower.isle

```diff
@@ -2572,6 +2572,249 @@
 (rule (lower (has_type (fits_in_64 ty) (uadd_overflow_trap a b tc)))
       (trap_if_overflow (add_with_flags_paired ty a b) tc))
 
+;;;; Helpers for `*_overflow` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; when true, sign extend, otherwise zero extend
+(decl put_in_reg_ext32 (Value bool) Reg)
+(rule (put_in_reg_ext32 val $true)
+      (put_in_reg_sext32 val))
+(rule (put_in_reg_ext32 val $false)
+      (put_in_reg_zext32 val))
+
+;; For values smaller than a register, we emit a normal op with both arguments
+;; zero/sign extended. We then check if the output is the same as itself zero/sign extended.
+(decl overflow_op_small (Type Value Value bool ALUOp) InstOutput)
+(rule (overflow_op_small ty a b sext alu_op)
+      (let ((extend ExtendOp (lower_extend_op ty sext))
+
+            ;; Instead of emitting two `{u,s}xt{b,h}` we do one as an instruction and
+            ;; the other as an extend operation in the alu_op.
+            ;;
+            ;; uxtb    a_ext, a
+            ;; alu_op  out, a_ext, b, {u,s}xtb
+            ;; cmp     out, out, {u,s}xtb
+            ;; cset    out_of, ne
+            (a_ext Reg (put_in_reg_ext32 a sext))
+            (out Reg (alu_rrr_extend alu_op ty a_ext b extend))
+            (out_of Reg (with_flags_reg 
+                  (cmp_extend (OperandSize.Size32) out out extend)
+                  (cset (Cond.Ne)))))
+      (output_pair
+            (value_reg out)
+            (value_reg out_of))))
+
+;; For register sized op's we just emit a op+cset, without further masking.
+;;
+;; op out, a, b
+;; cset out_of, cond
+(decl overflow_op_normal (Type Value Value ALUOp Cond) InstOutput)
```

Comment:
It's somewhat surprising here that we take an arbitrary `Cond`, given that the helper name is specifically "overflow"-related; could we document which `Cond`s we expect, or if it's only one in practice, just name it directly?

Reply:
I added a description which conditions are used.